## Correlation and Variance Inflation, Trip Distance (commute)

This is a notebook to compute variance inflation factors and correlation coefficients between urban form and selected other features for models of commute trip distance in 19 European cities.

In general, we try to avoid VIF>5, and correlation factors > 0.7 (absolute value)

In [1]:
# script to model avergage trip distances for commute trips in all cities
# last update Peter Berrill Nov 20 2023

# load in required packages
import numpy as np
import pandas as pd
import shap
import re
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score, RepeatedStratifiedKFold, cross_validate, GroupKFold, StratifiedGroupKFold, RepeatedKFold, StratifiedKFold, GridSearchCV, KFold
from sklearn import metrics, linear_model
from xgboost import XGBClassifier, XGBRegressor
import os
import sys
import matplotlib.pyplot as plt
import pickle
import statsmodels.formula.api as smf
from datetime import datetime
from statsmodels.stats.outliers_influence import variance_inflation_factor
from statsmodels.tools.tools import add_constant
import seaborn as sns

cities_all=['Berlin','Dresden','Düsseldorf','Frankfurt am Main','Kassel','Leipzig','Magdeburg','Potsdam','Clermont','Dijon','Lille','Lyon','Montpellier','Nantes','Nimes','Paris','Toulouse','Madrid','Wien','France_other','Germany_other']
countries=['Germany','Germany','Germany','Germany','Germany','Germany','Germany','Germany','France','France','France','France','France','France','France','France','France','Spain','Austria','France','Germany']


pandas.Int64Index is deprecated and will be removed from pandas in a future version. Use pandas.Index with the appropriate dtype instead.


In [3]:
def vif_corr(city):
    country=countries[cities_all.index(city)]
    print(city, country)
    if city=='Germany_other':
        city0='Dresden'
        df0=pd.read_csv('../outputs/Combined/' + city0 + '_UF.csv')
        df0.loc[(df0['Training'].isin(['Apprenticeship/Business','Craftsman/Technical'])) & (df0['Education']!='University'),'Education']='Apprenticeship'
        df0=df0.loc[:,['HH_P_WNR','HH_PNR', 'HHNR','Ori_geocode', 'Des_geocode','Res_geocode', 
                    'Trip_Time', 'Season','Trip_Purpose_Agg','HHSize',
                    'Sex', 'Occupation', 'Education','Age',
                    #'PopDensity_res','BuildDensity_res',
                    'UrbPopDensity_res', 'UrbBuildDensity_res','DistSubcenter_res', 'DistCenter_res',
                    'IntersecDensity_res', 'street_length_res', 'LU_UrbFab_res',#'bike_lane_share_res',
                    'LU_Comm_res' ,'Trip_Distance']]
        df0['City']=city0
        df_all=df0.copy()

        cities0=['Leipzig','Magdeburg','Potsdam','Frankfurt am Main','Düsseldorf','Kassel']
        for city1 in cities0:
                df1=pd.read_csv('../outputs/Combined/' + city1 + '_UF.csv')
                df1.loc[(df1['Training'].isin(['Apprenticeship/Business','Craftsman/Technical'])) & (df1['Education']!='University'),'Education']='Apprenticeship'
                df1=df1.loc[:,['HH_P_WNR','HH_PNR', 'HHNR','Ori_geocode', 'Des_geocode','Res_geocode', 
                            'Trip_Time', 'Season','Trip_Purpose_Agg','HHSize',
                            'Sex', 'Occupation', 'Education','Age',
                            #'PopDensity_res','BuildDensity_res',
                            'UrbPopDensity_res', 'UrbBuildDensity_res','DistSubcenter_res', 'DistCenter_res',
                            'IntersecDensity_res', 'street_length_res', 'LU_UrbFab_res',#'bike_lane_share_res',
                            'LU_Comm_res','Trip_Distance']]
                df1['City']=city1
                if len(df1.columns==df_all.columns):
                       df_all=pd.concat([df_all,df1])
        df_all['HHNR']=df_all['City']+'_'+df_all['HHNR'].astype(int).astype(str)
        df_all['HH_PNR']=df_all['City']+'_'+df_all['HH_PNR'].astype(int).astype(str)
        df_all['HH_P_WNR']=df_all['City']+'_'+df_all['HH_P_WNR'].astype(str)
        df_all.drop(columns='City',inplace=True)
        df_UF=df_all.copy()
    elif city=='France_other':
        city0='Clermont'
        df0=pd.read_csv('../outputs/Combined/' + city0 + '_UF.csv')
        df0=df0.loc[:,['HH_P_WNR','HH_PNR', 'HHNR','Ori_geocode', 'Des_geocode','Res_geocode', 
                    'Trip_Time', 'Season','Trip_Purpose_Agg','HHSize',
                    'Sex', 'Occupation', 'Education','Age',
                    #'PopDensity_res','BuildDensity_res',
                    'UrbPopDensity_res', 'UrbBuildDensity_res','DistSubcenter_res', 'DistCenter_res',
                    'IntersecDensity_res', 'street_length_res', 'LU_UrbFab_res',#'bike_lane_share_res',
                    'LU_Comm_res', 'Trip_Distance']]
        df0['City']=city0
        df_all=df0.copy()

        cities0=['Toulouse','Montpellier','Lyon','Nantes','Nimes','Lille','Dijon']
        for city1 in cities0:
                df1=pd.read_csv('../outputs/Combined/' + city1 + '_UF.csv')
                df1=df1.loc[:,['HH_P_WNR','HH_PNR', 'HHNR','Ori_geocode', 'Des_geocode','Res_geocode', 
                            'Trip_Time', 'Season','Trip_Purpose_Agg','HHSize',
                            'Sex', 'Occupation', 'Education','Age',
                            #'PopDensity_res','BuildDensity_res',
                            'UrbPopDensity_res', 'UrbBuildDensity_res','DistSubcenter_res', 'DistCenter_res',
                            'IntersecDensity_res', 'street_length_res', 'LU_UrbFab_res',#'bike_lane_share_res',
                            'LU_Comm_res', 'Trip_Distance']]
                df1['City']=city1
                if len(df1.columns==df_all.columns):
                       df_all=pd.concat([df_all,df1])
        df_all['HHNR']=df_all['City']+'_'+df_all['HHNR'].astype(int).astype(str)
        df_all['HH_PNR']=df_all['City']+'_'+df_all['HH_PNR'].astype(int).astype(str)
        df_all['HH_P_WNR']=df_all['City']+'_'+df_all['HH_P_WNR'].astype(str)
        df_all.drop(columns='City',inplace=True)
        df_UF=df_all.copy()
    else:
            df=pd.read_csv('../outputs/Combined/' + city + '_UF.csv',dtype={'Ori_geocode': str, 'Des_geocode': str,'Res_geocode': str })
            if country=='Germany':
                df.loc[(df['Training'].isin(['Apprenticeship/Business','Craftsman/Technical'])) & (df['Education']!='University'),'Education']='Apprenticeship'
            df_UF=df.loc[:,['HH_P_WNR','HH_PNR', 'HHNR','Ori_geocode', 'Des_geocode','Res_geocode', 
                            'Trip_Time', 'Season','Trip_Purpose_Agg','HHSize',
                            'Sex', 'Occupation', 'Education','Age',
                            #'PopDensity_res','BuildDensity_res',
                            'UrbPopDensity_res', 'UrbBuildDensity_res','DistSubcenter_res', 'DistCenter_res',
                            'IntersecDensity_res', 'street_length_res', 'LU_UrbFab_res',#'bike_lane_share_res',
                            'LU_Comm_res', 'Trip_Distance']]
    # restrict to trips between home and work (commuting trips)        
    df_UF=df_UF.loc[df_UF['Trip_Purpose_Agg']=='Home↔Work',]
    df_UF.drop(columns='Trip_Purpose_Agg',inplace=True)
    # restrict to those in employment
    df_UF=df_UF.loc[df_UF['Occupation'].isin(['Trainee','Employed_FullTime','Employed_PartTime','Employed']),]

    Edu_dict={'University':'University','Secondary':'Secondary','Secondary+BAC':'Secondary','Secondary+Matura':'Secondary',
          'Apprenticeship':'Apprenticeship',
          'Elementary':'Primary/None','Pre-School':'Primary/None','No diploma yet':'Primary/None','Unknown':'Primary/None','Other':'Primary/None'}

    df_UF['Education']=df_UF['Education'].map(Edu_dict)
    if city in ['Clermont','Nimes']:
          df_UF.loc[df_UF['Education']=='Apprenticeship','Education']='Secondary'

    df=df_UF.dropna()
    df['Sex']=df['Sex']-1 # change from [1,2] to [0,1], for plotting purposes
    df=df.loc[df['UrbBuildDensity_res']<1e8,]   # remove high building density outliers (For Leipzig)

    # identify the feature columns
    N_non_feature=6 # how many non-features are at the start of the df
    cols=df.columns
    newcols=(df.columns[:N_non_feature].tolist()) + ('FeatureD' +'_'+ cols[N_non_feature:-1]).tolist() + (df.columns[-1:].tolist())
    # change column names
    df.set_axis(newcols,axis=1,inplace=True)
    df = df.reset_index(drop=True)
    df0=df.copy()

    # convert  all categorical variables to dummies
    df_Cat=df.select_dtypes('object')[[col for col in df.select_dtypes('object').columns if "FeatureD" in col]]
    for col in df_Cat:
        dum=pd.get_dummies(df[[col]])
        df = pd.concat([df, dum], axis = 1)
        # remove the original categorical columns
    df.drop(df_Cat.columns.tolist(),axis=1,inplace=True)

    X=df[[col for col in df.columns if "FeatureD" in col]]
    return X

In [4]:
X=vif_corr('Berlin')
X_disp=[re.sub('FeatureD_','', x) for x in X.columns]
X.columns=X_disp
X = X[X.columns.drop(list(X.filter(regex='Education')))]
X = X[X.columns.drop(list(X.filter(regex='Season')))]
X = X[X.columns.drop(list(X.filter(regex='Trip_Time')))]
X = X[X.columns.drop(list(X.filter(regex='Occupation')))]
X_selc = add_constant(X)
vif=pd.Series([variance_inflation_factor(X_selc.values, i) for i in range(X_selc.shape[1])], index=X_selc.columns)
vif=vif.to_frame().reset_index()
vif.columns=['Feature','VIF']
corr = X.corr()
print(vif)
corr.style.format(precision=3).background_gradient(cmap='coolwarm')

Berlin Germany
                Feature         VIF
0                 const  227.031972
1                HHSize    1.049988
2                   Sex    1.004526
3                   Age    1.084612
4     UrbPopDensity_res    4.185069
5   UrbBuildDensity_res    4.025003
6     DistSubcenter_res    1.408009
7        DistCenter_res    3.046970
8   IntersecDensity_res    2.917028
9     street_length_res    1.935022
10        LU_UrbFab_res    1.930529
11          LU_Comm_res    1.959696


In Berlin, VIFs are acceptable, although built-up density has a high correlation (0.768) with population density.

Repeating without built-up density leads to notably lower VIFs for population density.

In [5]:
# Berlin better VIF w/o bld dens
X.drop(columns='UrbBuildDensity_res',inplace=True)
X_selc = add_constant(X)
vif=pd.Series([variance_inflation_factor(X_selc.values, i) for i in range(X_selc.shape[1])], index=X_selc.columns)
vif=vif.to_frame().reset_index()
vif.columns=['Feature','VIF']
corr = X.corr()
print(vif)
corr.style.format(precision=3).background_gradient(cmap='coolwarm')

                Feature         VIF
0                 const  226.606180
1                HHSize    1.049856
2                   Sex    1.004491
3                   Age    1.082793
4     UrbPopDensity_res    2.870375
5     DistSubcenter_res    1.407901
6        DistCenter_res    2.872738
7   IntersecDensity_res    2.881996
8     street_length_res    1.932668
9         LU_UrbFab_res    1.929822
10          LU_Comm_res    1.599221


In [6]:
X=vif_corr('Paris')
X_disp=[re.sub('FeatureD_','', x) for x in X.columns]
X.columns=X_disp
X = X[X.columns.drop(list(X.filter(regex='Education')))]
X = X[X.columns.drop(list(X.filter(regex='Season')))]
X = X[X.columns.drop(list(X.filter(regex='Trip_Time')))]
X = X[X.columns.drop(list(X.filter(regex='Occupation')))]
X_selc = add_constant(X)
vif=pd.Series([variance_inflation_factor(X_selc.values, i) for i in range(X_selc.shape[1])], index=X_selc.columns)
vif=vif.to_frame().reset_index()
vif.columns=['Feature','VIF']
corr = X.corr()
print(vif)
corr.style.format(precision=3).background_gradient(cmap='coolwarm')

Paris France
                Feature         VIF
0                 const  376.262980
1                HHSize    1.046552
2                   Sex    1.007918
3                   Age    1.009086
4     UrbPopDensity_res    5.519999
5   UrbBuildDensity_res    4.168286
6     DistSubcenter_res    1.489145
7        DistCenter_res    3.615204
8   IntersecDensity_res    4.151349
9     street_length_res    2.128021
10        LU_UrbFab_res    2.504173
11          LU_Comm_res    1.811359


In Paris, VIFs is high for population density, and other density measures.

Repeating without built-up density leads to notably lower VIFs for population density, it is still a bit high for intersection density (4.1) but acceptable.

Correlation remains high between population density and distance to center (0.772) but we retain these variables due to their high importance.

In [7]:
# Paris better VIF w/o bld dens
X.drop(columns='UrbBuildDensity_res',inplace=True)
X_selc = add_constant(X)
vif=pd.Series([variance_inflation_factor(X_selc.values, i) for i in range(X_selc.shape[1])], index=X_selc.columns)
vif=vif.to_frame().reset_index()
vif.columns=['Feature','VIF']
corr = X.corr()
print(vif)
corr.style.format(precision=3).background_gradient(cmap='coolwarm')

                Feature         VIF
0                 const  374.397246
1                HHSize    1.045705
2                   Sex    1.007918
3                   Age    1.008345
4     UrbPopDensity_res    3.450744
5     DistSubcenter_res    1.476201
6        DistCenter_res    3.407399
7   IntersecDensity_res    4.149978
8     street_length_res    2.127007
9         LU_UrbFab_res    2.485582
10          LU_Comm_res    1.691672


In [8]:
X=vif_corr('Madrid')
X_disp=[re.sub('FeatureD_','', x) for x in X.columns]
X.columns=X_disp
X = X[X.columns.drop(list(X.filter(regex='Education')))]
X = X[X.columns.drop(list(X.filter(regex='Season')))]
X = X[X.columns.drop(list(X.filter(regex='Trip_Time')))]
X = X[X.columns.drop(list(X.filter(regex='Occupation')))]
X_selc = add_constant(X)
vif=pd.Series([variance_inflation_factor(X_selc.values, i) for i in range(X_selc.shape[1])], index=X_selc.columns)
vif=vif.to_frame().reset_index()
vif.columns=['Feature','VIF']
corr = X.corr()
print(vif)
corr.style.format(precision=3).background_gradient(cmap='coolwarm')

Madrid Spain



A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


                Feature         VIF
0                 const  222.577266
1                HHSize    1.029913
2                   Sex    1.001847
3                   Age    1.017704
4     UrbPopDensity_res    2.380048
5   UrbBuildDensity_res    1.248022
6     DistSubcenter_res    1.282889
7        DistCenter_res    1.594868
8   IntersecDensity_res    2.718169
9     street_length_res    2.072874
10        LU_UrbFab_res    2.098204
11          LU_Comm_res    1.311277


No concerns surrounding VIF or correlations in Madrid. 
However, the highest correlation factor between population density and urban fabric land use (0.658) does coincide with suspicious coefficientsin the linear regression models (positive coefficient for pop density and a very large negative coeff for LU UF).
Therefore, we will drop urban fabric land use, for more reliable regression coefficient estimates.

VIF for pop density than drops from 2.38 to 1.65.

In [9]:
# we drop LU Urb Fab from this model. Similar for DE, other
X.drop(columns=['LU_UrbFab_res'],inplace=True)
X_selc = add_constant(X)
vif=pd.Series([variance_inflation_factor(X_selc.values, i) for i in range(X_selc.shape[1])], index=X_selc.columns)
vif=vif.to_frame().reset_index()
vif.columns=['Feature','VIF']
corr = X.corr()
print(vif)
corr.style.format(precision=3).background_gradient(cmap='coolwarm')

                Feature         VIF
0                 const  216.650848
1                HHSize    1.028034
2                   Sex    1.001762
3                   Age    1.013341
4     UrbPopDensity_res    1.653727
5   UrbBuildDensity_res    1.246587
6     DistSubcenter_res    1.272258
7        DistCenter_res    1.594677
8   IntersecDensity_res    2.712061
9     street_length_res    2.069157
10          LU_Comm_res    1.148918


In [11]:
X=vif_corr('Wien')
X_disp=[re.sub('FeatureD_','', x) for x in X.columns]
X.columns=X_disp
X = X[X.columns.drop(list(X.filter(regex='Education')))]
X = X[X.columns.drop(list(X.filter(regex='Season')))]
X = X[X.columns.drop(list(X.filter(regex='Trip_Time')))]
X = X[X.columns.drop(list(X.filter(regex='Occupation')))]
X_selc = add_constant(X)
vif=pd.Series([variance_inflation_factor(X_selc.values, i) for i in range(X_selc.shape[1])], index=X_selc.columns)
vif=vif.to_frame().reset_index()
vif.columns=['Feature','VIF']
corr = X.corr()
print(vif)
corr.style.format(precision=3).background_gradient(cmap='coolwarm')

Wien Austria
                Feature         VIF
0                 const  334.483550
1                HHSize    1.063261
2                   Sex    1.022925
3                   Age    1.015436
4     UrbPopDensity_res    7.474097
5   UrbBuildDensity_res    2.985189
6     DistSubcenter_res    2.231792
7        DistCenter_res    4.596859
8   IntersecDensity_res    9.262111
9     street_length_res    5.719510
10        LU_UrbFab_res    5.739174
11          LU_Comm_res    1.792086


Vienna has several problematic correlations, the highest VIF is for intersection density, which has corr of 0.9 with street lenght. LU urban fabric also has high VIF and correlation with pop density.
After dropping intersection density and LU urban fabric VIFs are much lower, distance to center remains somewhat high at 4.0, but acceptable. 
Correlation is high (0.743) between population density and distance to center, but we keep both features due to their importance.

In [12]:
X.drop(columns=['IntersecDensity_res','LU_UrbFab_res'],inplace=True)
X = X[X.columns.drop(list(X.filter(regex='Occupation')))]
X_selc = add_constant(X)
vif=pd.Series([variance_inflation_factor(X_selc.values, i) for i in range(X_selc.shape[1])], index=X_selc.columns)
vif=vif.to_frame().reset_index()
vif.columns=['Feature','VIF']
corr = X.corr()
print(vif)
corr.style.format(precision=3).background_gradient(cmap='coolwarm')

               Feature         VIF
0                const  112.804796
1               HHSize    1.061433
2                  Sex    1.022084
3                  Age    1.015182
4    UrbPopDensity_res    3.381877
5  UrbBuildDensity_res    2.679027
6    DistSubcenter_res    1.702374
7       DistCenter_res    4.022401
8    street_length_res    2.116933
9          LU_Comm_res    1.636831


In [16]:
X=vif_corr('Germany_other')
X_disp=[re.sub('FeatureD_','', x) for x in X.columns]
X.columns=X_disp
X = X[X.columns.drop(list(X.filter(regex='Education')))]
X = X[X.columns.drop(list(X.filter(regex='Season')))]
X = X[X.columns.drop(list(X.filter(regex='Trip_Time')))]
X = X[X.columns.drop(list(X.filter(regex='Occupation')))]
X_selc = add_constant(X)
vif=pd.Series([variance_inflation_factor(X_selc.values, i) for i in range(X_selc.shape[1])], index=X_selc.columns)
vif=vif.to_frame().reset_index()
vif.columns=['Feature','VIF']
corr = X.corr()
print(vif)
corr.style.format(precision=3).background_gradient(cmap='coolwarm')

Germany_other Germany


Columns (4,5) have mixed types. Specify dtype option on import or set low_memory=False.


                Feature         VIF
0                 const  185.006054
1                HHSize    1.047870
2                   Sex    1.005770
3                   Age    1.042088
4     UrbPopDensity_res    3.263240
5   UrbBuildDensity_res    2.040155
6     DistSubcenter_res    1.199650
7        DistCenter_res    1.597555
8   IntersecDensity_res    3.866636
9     street_length_res    1.921117
10        LU_UrbFab_res    3.261613
11          LU_Comm_res    1.472240


VIFs are acceptable low for rest of Germany. Some high correlation between urban fabric land use with pop density (0.772) and intersection density (0.739). Intersection density also quite highly correlated with street length and population density. 

Dropping urban fabric land use and intersection density leads to all very low VIFs and correlations.

In [17]:
X.drop(columns=['LU_UrbFab_res','IntersecDensity_res'],inplace=True)
X_selc = add_constant(X)
vif=pd.Series([variance_inflation_factor(X_selc.values, i) for i in range(X_selc.shape[1])], index=X_selc.columns)
vif=vif.to_frame().reset_index()
vif.columns=['Feature','VIF']
corr = X.corr()
print(vif)
corr.style.format(precision=3).background_gradient(cmap='coolwarm')

               Feature         VIF
0                const  101.048895
1               HHSize    1.046360
2                  Sex    1.005740
3                  Age    1.041378
4    UrbPopDensity_res    1.790977
5  UrbBuildDensity_res    1.985419
6    DistSubcenter_res    1.198307
7       DistCenter_res    1.335819
8    street_length_res    1.404958
9          LU_Comm_res    1.459573


In [21]:
X=vif_corr('France_other')
X_disp=[re.sub('FeatureD_','', x) for x in X.columns]
X.columns=X_disp
X = X[X.columns.drop(list(X.filter(regex='Education')))]
X = X[X.columns.drop(list(X.filter(regex='Season')))]
X = X[X.columns.drop(list(X.filter(regex='Trip_Time')))]
X = X[X.columns.drop(list(X.filter(regex='Occupation')))]
X_selc = add_constant(X)
vif=pd.Series([variance_inflation_factor(X_selc.values, i) for i in range(X_selc.shape[1])], index=X_selc.columns)
vif=vif.to_frame().reset_index()
vif.columns=['Feature','VIF']
corr = X.corr()
print(vif)
corr.style.format(precision=3).background_gradient(cmap='coolwarm')

France_other France



A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


                Feature         VIF
0                 const  118.898692
1                HHSize    1.055082
2                   Sex    1.009441
3                   Age    1.032519
4     UrbPopDensity_res    3.345826
5   UrbBuildDensity_res    2.868368
6     DistSubcenter_res    1.535069
7        DistCenter_res    1.838679
8   IntersecDensity_res    4.004131
9     street_length_res    1.839450
10        LU_UrbFab_res    1.959435
11          LU_Comm_res    1.381690


Rest of France has acceptably low VIFs and correlations, highest VIF is 4.0 for intersection density, which has moderatly high correlation with population density.
Dropping intersection density would also improve comparability between rest of France and rest of Germany, and would maintain consistency between dist agg and dist commute models for rest of France.

In [22]:
X.drop(columns=['IntersecDensity_res'],inplace=True)
X_selc = add_constant(X)
vif=pd.Series([variance_inflation_factor(X_selc.values, i) for i in range(X_selc.shape[1])], index=X_selc.columns)
vif=vif.to_frame().reset_index()
vif.columns=['Feature','VIF']
corr = X.corr()
print(vif)
corr.style.format(precision=3).background_gradient(cmap='coolwarm')

                Feature        VIF
0                 const  72.721763
1                HHSize   1.053835
2                   Sex   1.009433
3                   Age   1.032183
4     UrbPopDensity_res   3.233387
5   UrbBuildDensity_res   2.624639
6     DistSubcenter_res   1.526662
7        DistCenter_res   1.667172
8     street_length_res   1.282810
9         LU_UrbFab_res   1.818256
10          LU_Comm_res   1.380271


In [23]:
# all city model
city='Berlin'
country=countries[cities_all.index(city)]
print(city, country)

city0='Berlin'
df0=pd.read_csv('../outputs/Combined/' + city0 + '_UF.csv')
# add to make consistent with definition of apprenticeship within Education in other countries
df0.loc[(df0['Training'].isin(['Apprenticeship/Business','Craftsman/Technical'])) & (df0['Education']!='University'),'Education']='Apprenticeship'
df0=df0.loc[:,['HH_P_WNR','HH_PNR', 'HHNR','Ori_geocode', 'Des_geocode','Res_geocode', 
            'Trip_Time', 'Season','Trip_Purpose_Agg','HHSize',
            'Sex', 'Occupation', 'Education','Age',
            #'PopDensity_res','BuildDensity_res',
            'UrbPopDensity_res', 'UrbBuildDensity_res','DistSubcenter_res', 'DistCenter_res',
            'IntersecDensity_res', 'street_length_res', 'LU_UrbFab_res',#'bike_lane_share_res',
            'LU_Comm_res' ,'Trip_Distance']]
df0['City']=city0
df0['Country']='Germany'
df_all=df0.copy()

cities0=['Dresden', 'Leipzig','Magdeburg','Potsdam','Frankfurt am Main','Düsseldorf','Kassel']
for city1 in cities0:
        print(city1)
        df1=pd.read_csv('../outputs/Combined/' + city1 + '_UF.csv')
        df1.loc[(df1['Training'].isin(['Apprenticeship/Business','Craftsman/Technical'])) & (df0['Education']!='University'),'Education']='Apprenticeship'
        df1=df1.loc[:,['HH_P_WNR','HH_PNR', 'HHNR','Ori_geocode', 'Des_geocode','Res_geocode', 
                    'Trip_Time', 'Season','Trip_Purpose_Agg','HHSize',
                    'Sex', 'Occupation', 'Education','Age',
                    #'PopDensity_res','BuildDensity_res',
                    'UrbPopDensity_res', 'UrbBuildDensity_res','DistSubcenter_res', 'DistCenter_res',
                    'IntersecDensity_res', 'street_length_res', 'LU_UrbFab_res',#'bike_lane_share_res',
                    'LU_Comm_res','Trip_Distance']]
        df1['City']=city1
        df1['Country']='Germany'
        if len(df1.columns==df_all.columns):
                df_all=pd.concat([df_all,df1])
                print(city1, 'added.')
                #print(len(df_all), 'rows in the combined dataframe')
df_all['HHNR']=df_all['City']+'_'+df_all['HHNR'].astype(int).astype(str)
df_all['HH_PNR']=df_all['City']+'_'+df_all['HH_PNR'].astype(int).astype(str)
df_all['HH_P_WNR']=df_all['City']+'_'+df_all['HH_P_WNR'].astype(str)
df_all.drop(columns='City',inplace=True)
df_DE=df_all.copy()
city0='Clermont'
df0=pd.read_csv('../outputs/Combined/' + city0 + '_UF.csv')
df0=df0.loc[:,['HH_P_WNR','HH_PNR', 'HHNR','Ori_geocode', 'Des_geocode','Res_geocode', 
            'Trip_Time', 'Season','Trip_Purpose_Agg','HHSize',
            'Sex', 'Occupation', 'Education','Age',
            #'PopDensity_res','BuildDensity_res',
            'UrbPopDensity_res', 'UrbBuildDensity_res','DistSubcenter_res', 'DistCenter_res',
            'IntersecDensity_res', 'street_length_res', 'LU_UrbFab_res',#'bike_lane_share_res',
            'LU_Comm_res', 'Trip_Distance']]
df0['City']=city0
df0['Country']='France'
df_all=df0.copy()

cities0=['Toulouse','Montpellier','Lyon','Nantes','Nimes','Lille','Dijon','Paris']
for city1 in cities0:
        print(city1)
        df1=pd.read_csv('../outputs/Combined/' + city1 + '_UF.csv')
        df1=df1.loc[:,['HH_P_WNR','HH_PNR', 'HHNR','Ori_geocode', 'Des_geocode','Res_geocode', 
                    'Trip_Time', 'Season','Trip_Purpose_Agg','HHSize',
                    'Sex', 'Occupation', 'Education','Age',
                    #'PopDensity_res','BuildDensity_res',
                    'UrbPopDensity_res', 'UrbBuildDensity_res','DistSubcenter_res', 'DistCenter_res',
                    'IntersecDensity_res', 'street_length_res', 'LU_UrbFab_res',#'bike_lane_share_res',
                    'LU_Comm_res', 'Trip_Distance']]
        df1['City']=city1
        df1['Country']='France'
        if len(df1.columns==df_all.columns):
                df_all=pd.concat([df_all,df1])
                print(city1, 'added.')
                #print(len(df_all), 'rows in the combined dataframe')
df_all['HHNR']=df_all['City']+'_'+df_all['HHNR'].astype(str)
df_all['HH_PNR']=df_all['City']+'_'+df_all['HH_PNR'].astype(str)
df_all['HH_P_WNR']=df_all['City']+'_'+df_all['HH_P_WNR'].astype(str)
df_all.drop(columns='City',inplace=True)
df_FR=df_all.copy()
df=pd.read_csv('../outputs/Combined/' + 'Madrid' + '_UF.csv',dtype={'Ori_geocode': str, 'Des_geocode': str,'Res_geocode': str })
df_UF=df.loc[:,['HH_P_WNR','HH_PNR', 'HHNR','Ori_geocode', 'Des_geocode','Res_geocode', 
                'Trip_Time', 'Season','Trip_Purpose_Agg','HHSize',
                'Sex', 'Occupation', 'Education','Age',
                #'PopDensity_res','BuildDensity_res',
                'UrbPopDensity_res', 'UrbBuildDensity_res','DistSubcenter_res', 'DistCenter_res',
                'IntersecDensity_res', 'street_length_res', 'LU_UrbFab_res',#'bike_lane_share_res',
                'LU_Comm_res', 'Trip_Distance']]
df_UF['City']='Madrid'
df_UF['Country']='Spain'
df_UF['HHNR']=df_UF['City']+'_'+df_UF['HHNR'].astype(str)
df_UF['HH_PNR']=df_UF['City']+'_'+df_UF['HH_PNR'].astype(str)
df_UF['HH_P_WNR']=df_UF['City']+'_'+df_UF['HH_P_WNR'].astype(str)
df_UF.drop(columns='City',inplace=True)
df_Madrid=df_UF.copy()
df=pd.read_csv('../outputs/Combined/' + 'Wien' + '_UF.csv',dtype={'Ori_geocode': str, 'Des_geocode': str,'Res_geocode': str })
df_UF=df.loc[:,['HH_P_WNR','HH_PNR', 'HHNR','Ori_geocode', 'Des_geocode','Res_geocode', 
                'Trip_Time', 'Season','Trip_Purpose_Agg','HHSize',
                'Sex', 'Occupation', 'Education','Age',
                #'PopDensity_res','BuildDensity_res',
                'UrbPopDensity_res', 'UrbBuildDensity_res','DistSubcenter_res', 'DistCenter_res',
                'IntersecDensity_res', 'street_length_res', 'LU_UrbFab_res',#'bike_lane_share_res',
                'LU_Comm_res', 'Trip_Distance']]
df_UF['City']='Wien'
df_UF['Country']='Austria'
df_UF['HHNR']=df_UF['City']+'_'+df_UF['HHNR'].astype(str)
df_UF['HH_PNR']=df_UF['City']+'_'+df_UF['HH_PNR'].astype(str)
df_UF['HH_P_WNR']=df_UF['City']+'_'+df_UF['HH_P_WNR'].astype(str)
df_UF.drop(columns='City',inplace=True)
df_Wien=df_UF.copy()
df_UF=pd.concat([df_DE,df_FR,df_Madrid,df_Wien],ignore_index=True)
df_UF['Trip_Purpose_Agg'].value_counts()

Berlin Germany


Columns (4) have mixed types. Specify dtype option on import or set low_memory=False.


Dresden
Dresden added.
Leipzig
Leipzig added.
Magdeburg
Magdeburg added.
Potsdam
Potsdam added.
Frankfurt am Main
Frankfurt am Main added.
Düsseldorf
Düsseldorf added.
Kassel
Kassel added.


Columns (4,5) have mixed types. Specify dtype option on import or set low_memory=False.


Toulouse
Toulouse added.
Montpellier
Montpellier added.
Lyon
Lyon added.
Nantes
Nantes added.
Nimes
Nimes added.
Lille
Lille added.
Dijon
Dijon added.
Paris
Paris added.


Home↔Leisure      156611
Other             132906
Home↔Work         130623
Home↔Shopping     108008
Home↔School        83984
Home↔Companion     62002
Name: Trip_Purpose_Agg, dtype: int64

In [5]:
df_UF=pd.concat([df_DE,df_FR,df_Madrid,df_Wien],ignore_index=True)
df_UF=df_UF.loc[df_UF['Trip_Purpose_Agg']=='Home↔Work',]
df_UF.drop(columns='Trip_Purpose_Agg',inplace=True)

Occ_dict={'Employed_FullTime':'Employed','Employed_PartTime':'Employed','Employed':'Employed','Trainee':'Employed',
          'Student_School':'Student_School','Student_3rdLevel':'Student_3rdLevel','Pre-School':'Pre-School','Retired':'Retired',
          'Unemployed':'Unemployed/Other','Other':'Unemployed/Other','Home_Partner':'Unemployed/Other'}
Edu_dict={'University':'University','Secondary':'Secondary','Secondary+BAC':'Secondary','Secondary+Matura':'Secondary',
          'Apprenticeship':'Apprenticeship',
          'Elementary':'Primary/None','Pre-School':'Primary/None','No diploma yet':'Primary/None','Unknown':'Primary/None','Other':'Primary/None'}

df_UF['Occupation']=df_UF['Occupation'].map(Occ_dict)
df_UF['Education']=df_UF['Education'].map(Edu_dict)

df_UF=pd.concat([df_UF.drop(columns='Trip_Distance'),df_UF['Trip_Distance']],axis=1)

df=df_UF.dropna()
df['Sex']=df['Sex']-1 # change from [1,2] to [0,1], for plotting purposes
df=df.loc[df['UrbBuildDensity_res']<1e8,]   # remove high building density outliers (For Leipzig)

# identify the feature columns
N_non_feature=6 # how many non-features are at the start of the df
cols=df.columns
newcols=(df.columns[:N_non_feature].tolist()) + ('FeatureD' +'_'+ cols[N_non_feature:-1]).tolist() + (df.columns[-1:].tolist())
# change column names
df.set_axis(newcols,axis=1,inplace=True)
df = df.reset_index(drop=True)
df0=df.copy()

# convert  all categorical variables to dummies
df_Cat=df.select_dtypes('object')[[col for col in df.select_dtypes('object').columns if "FeatureD" in col]]
for col in df_Cat:
    dum=pd.get_dummies(df[[col]])
    df = pd.concat([df, dum], axis = 1)
    # remove the original categorical columns
df.drop(df_Cat.columns.tolist(),axis=1,inplace=True)

X=df[[col for col in df.columns if "FeatureD" in col]]


A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy


No VIF or correlation concerns with all cities. Built-up and population density have high correlation (0.697) but VIFs are fine.

In [24]:
# no correlation concerns
X_disp=[re.sub('FeatureD_','', x) for x in X.columns]
X.columns=X_disp
X = X[X.columns.drop(list(X.filter(regex='Education')))]
X = X[X.columns.drop(list(X.filter(regex='Season')))]
X = X[X.columns.drop(list(X.filter(regex='Trip_Time')))]
X = X[X.columns.drop(list(X.filter(regex='Occupation')))]
X = X[X.columns.drop(list(X.filter(regex='Country')))]
X_selc = add_constant(X)
vif=pd.Series([variance_inflation_factor(X_selc.values, i) for i in range(X_selc.shape[1])], index=X_selc.columns)
vif=vif.to_frame().reset_index()
vif.columns=['Feature','VIF']
corr = X.corr()
print(vif)
corr.style.format(precision=3).background_gradient(cmap='coolwarm')

                Feature        VIF
0                 const  72.721763
1                HHSize   1.053835
2                   Sex   1.009433
3                   Age   1.032183
4     UrbPopDensity_res   3.233387
5   UrbBuildDensity_res   2.624639
6     DistSubcenter_res   1.526662
7        DistCenter_res   1.667172
8     street_length_res   1.282810
9         LU_UrbFab_res   1.818256
10          LU_Comm_res   1.380271
